src/cunumeric/fft/fft.cu

```diff
@@ -0,0 +1,351 @@
+                                      const Rect<3>& out_rect,
+                                      const Rect<3>& in_rect,
+                                      int axis,
+                                      fftDirection direction) {
+    bool is_inner_axis = (axis == 1);
+    if(is_inner_axis) {
+      // TODO: use PointInRectIterator<DIM>
+      auto num_slices = in_rect.hi[0] - in_rect.lo[0] + 1;
+      for(unsigned n = 0; n < num_slices; ++n){
+        const Point<3> offset = Point<3>(n, 0, 0);
+        CHECK_CUFFT(cufftXtExec(plan, (void*)in.ptr(offset), (void*)out.ptr(offset), (int)direction));
+      }
+    }
+    else {
+      const Point<3> zero   = Point<3>::ZEROES();
+      CHECK_CUFFT(cufftXtExec(plan, (void*)in.ptr(zero), (void*)out.ptr(zero), (int)direction));
+    }
+  }
+};
+
+// Perform the FFT operation as multiple 1D FFTs along the specified axes.
+// For now, it only supports up to 3D FFTs, but final plan is having support for
+// N-dimensional FFTs using this approach.
+template <int DIM, typename OUTPUT_TYPE, typename INPUT_TYPE>
+__host__ static inline void cufft_over_axes(AccessorWO<OUTPUT_TYPE, DIM> out,
+                                            AccessorRO<INPUT_TYPE, DIM> in,
+                                            const Rect<DIM>& out_rect,
+                                            const Rect<DIM>& in_rect,
+                                            std::vector<int64_t>& axes,
+                                            fftType type,
+                                            fftDirection direction)
+{
+    cudaStream_t stream;
+    CHECK_CUDA(cudaStreamCreateWithFlags(&stream, cudaStreamNonBlocking));
+
+    size_t workarea_size;
+    int n[DIM];
+    int inembed[DIM];
+    int onembed[DIM];
+
+    // Full volume dimensions / strides
+    const Point<DIM> zero   = Point<DIM>::ZEROES();
+    const Point<DIM> one    = Point<DIM>::ONES();
+    Point<DIM> fft_size_in  =  in_rect.hi -  in_rect.lo + one;
+    Point<DIM> fft_size_out = out_rect.hi - out_rect.lo + one;
+    size_t num_elements_in  = 1;
+    size_t num_elements_out = 1;
+    for(int i = 0; i < DIM; ++i) {
+      n[i]          = /*(type == fftType::FFT_R2C || type == fftType::FFT_D2Z) ? fft_size_in[i] :*/ fft_size_out[i];
+      inembed[i]    = fft_size_in[i];
+      onembed[i]    = fft_size_out[i];
+      num_elements_in  *= fft_size_in[i];
+      num_elements_out *= fft_size_out[i];
+    }
+
+    // Copy input to temporary buffer to perform FFTs one by one
+    DeferredBuffer<INPUT_TYPE, DIM> input_buffer(Rect<DIM>(zero, fft_size_in - one),
+                                                 Memory::GPU_FB_MEM,
+                                                 nullptr /*initial*/,
+                                                 128 /*alignment*/);
+    CHECK_CUDA(cudaMemcpyAsync(input_buffer.ptr(zero), in.ptr(zero), num_elements_in*sizeof(INPUT_TYPE), cudaMemcpyDeviceToDevice, stream));
```

Comment:
Any reason to not use `cudaMemcpyDefault` ?

Reply:
Isn't that less clear for developers?